src/System.Diagnostics.StackTrace.Symbols/src/System/Diagnostics/StackTrace/Symbols.cs

```diff
@@ -0,0 +1,204 @@
+                    reader = provider.GetMetadataReader();
+                }
+            }
+
+            return provider;
+        }
+
+        /// <summary>
+        /// Read the pdb file name and assembly version information from the PE file.
+        /// </summary>
+        /// <param name="assemblyPath">PE file path</param>
+        /// <param name="inMemoryAddress">in memory PE address</param>
+        /// <param name="inMemorySize">in memory PE size</param>
+        /// <param name="age">age</param>
+        /// <param name="guid">assembly guid</param>
+        /// <param name="stamp">time stamp</param>
+        /// <returns>pdb name or null</returns>
+        static unsafe string GetPdbPathFromPeStream(string assemblyPath, IntPtr inMemoryAddress, int inMemorySize, out int age, out Guid guid, out uint stamp)
+        {
+            try
+            {
+                if (inMemoryAddress != IntPtr.Zero && inMemorySize > 0)
+                {
+                    using (var peReader = new PEReader((byte*)inMemoryAddress.ToPointer(), inMemorySize))
+                    {
+                        foreach (var entry in peReader.ReadDebugDirectory())
+                        {
+                            if (entry.Type == DebugDirectoryEntryType.CodeView)
+                            {
+                                CodeViewDebugDirectoryData codeViewData = peReader.ReadCodeViewDebugDirectoryData(entry);
+
+                                stamp = entry.Stamp;
+                                age = codeViewData.Age;
+                                guid = codeViewData.Guid;
+
+                                string peDirectory = Path.GetDirectoryName(assemblyPath);
+                                return Path.Combine(peDirectory, Path.GetFileName(codeViewData.Path));
+                            }
+                        }
+                    }
+                }
+            }
+            catch
+            {
+            }
+
+            stamp = 0;
+            age = 0;
+            guid = new Guid();
+            return null;
+        }
+
+        /// <summary>
+        /// Returns true if the portable pdb id matches the guid and stamp.
+        /// </summary>
+        static bool IdEquals(ImmutableArray<byte> left, Guid rightGuid, uint rightStamp)
+        {
+            if (left.Length != 20)
+            {
+                // invalid id
+                return false;
+            }
+
+            byte[] guidBytes = rightGuid.ToByteArray();
+            for (int i = 0; i < guidBytes.Length; i++)
+            {
+                if (guidBytes[i] != left[i])
+                {
+                    return false;
+                }
+            }
+
+            byte[] stampBytes = BitConverter.GetBytes(rightStamp);
+            for (int i = 0; i < stampBytes.Length; i++)
+            {
+                if (stampBytes[i] != left[guidBytes.Length + i])
+                {
+                    return false;
+                }
+            }
```

Comment:
I don't know how important perf is here, but with a bit more work there would be ways to avoid these byte[] allocations if it is important.